### ISSUE TITLE
Feature Request: Popup to create pr/issues ..

### ISSUE THREAD

**Issue by tami5** [pwntester/octo.nvim#197]
## Issue Description

Type: *feature request*

### Describe what happened (or what feature you want)

I'd like to be simplify the way I create pr/issues and not having to wait and be promoted with a new buffer.

currently the steps to create an issue for me, goes as follows:

1. run `:Octo issue create` (I bind it to `<leader>gic`)
2. enter issue title in command line
3. wait till the create request finish
4. open a buffer (which I ONLY NEED TO WRITE DESCRIPTION)
5. write a message
6. choose labels (optional)

### Describe what you expected to happen

I'd like to have some sort of global create menu. Where in this menu is all the stuff that can be created using github cli or some git related stuff.

After choosing, say issue. a popup with something like:

```markdown

## Issue Title:
 |
## Issue description
<!-- You can Leave Empty-->
|<TAB to jump here>
```

Then on :w<CR> or in insert `<C-c>` the buffer will be parsed and sent to the api.

Then I'd like to be notified using something like notified using something like https://github.com/rcarriga/nvim-notify when the issue is created or when some network error happened (possibly cache the buffer for retry).

So the steps becomes as follows 

1. run `require'octo.telescope.menu'.create()`
2. select something that can be created using github api or git
3. small popup enter required fields in a markdown form.
4. hit `<C-c>` in insert mode or `<leader>w` in normal mode to save.
5. continue what I was doing
6. get notified when things the network request is done 💯  

### Anything else we need to know?

I'd something that involved parsing markdown form

first I write the following in empty popup buffer

```lua
{
    "#### Purpose",
    "",
    (opts.branch_desc or "") .. ".",
    "",
    "#### Details",
    "",
    first_commit_body,
 }
```

then after saving the buffer, I send the buffer to a function that does the following 

```lua
    local content = vim.api.nvim_buf_get_lines(win.bufnr, 0, -1, false)
    local title = content[2]
    local body = table.concat(vim.list_slice(content, 6, #content), "\n")
 --- ....
```

**Comment by pwntester** [comment]
Hey @tami5 

I had something similar on mind but with no menus/popups. What about:

- `Octo issue|pr create`
- new octo buffer with empty title and body is created
- user enters title and body (signcolumn indicators turn red)
- user saves the buffer (signcolumn indicators turn white) and vim.notify informs if the issue/pr was created successfully

**Comment by tami5** [comment]
well that is good enough @pwntester. I can later create a custom menu, and if it's stable enough maybe open a pr. 

the main pain point for me is having to wait until the request for creating an issue is created.

**Comment by pwntester** [comment]
Maybe #163 is enough for the menu thing. As for the popup, wouldnt you have to wait till the issue is created to show the octo buffer after entering the details on the popup? What Im suggesting is creating the octo buffer staight away and then create the issue on the background on the first save

**Comment by tami5** [comment]
> Maybe #163 is enough for the menu thing. As for the popup, wouldnt you have to wait till the issue is created to show the octo buffer after entering the details on the popup? What Im suggesting is creating the octo buffer staight away and then create the issue on the background on the first save

not sure if I understand what you're saying here but basically, yah no wait between creating the issue/pr and first save.

As for popup, it should just be a swap for regular splits and I think it would be an easy implementation. I'd like to see an option to only use popup for all octo buffer and be anle to toggle on/off them. But that's something for another feature request.

for now between entering title and description shouldn't be a network request. 

As for UI, one place to write title and description

**Comment by disrupted** [comment]
I've made an interactive form for _Create PR_ using [nui-components.nvim](https://github.com/grapp-dev/nui-components.nvim) to see how that could look like.

Here's a little demo:

https://github.com/user-attachments/assets/2057462d-593a-4ad7-b600-05b9edd291ec

At the moment this lives in my dotfiles. https://github.com/disrupted/dotfiles/blob/7666dc7644e59c5a9d2fb2911fb41a34467b09ca/.config/nvim/lua/conf/octo.lua#L277

I am contemplating whether this could be extracted into a plugin or perhaps upstreamed. Would you be interested to include something like this or is it out of scope for Octo?

**Comment by wd60622** [comment]
Really cool! @disrupted Looks solid

Maybe we can have a custom `collect_pr_input` which can be overridden. Not sure if forcing users to have this dependency is immediate route as it would require people to change their configs. 

As I've mentioned in other issues, the current `Octo pr create` logic is pretty involved and maybe needs a refactor to expose.

Note, that users can override any command with the `commands` section of the configuration. For example, 

```lua
require("octo").setup {
  commands = {
    pr = {
      create = function()
        require("my-custom-module").pr_create()
      end,
    },
  },
}
```

**Comment by disrupted** [comment]
> Not sure if forcing users to have this dependency is immediate route as it would require people to change their configs.

yep, that's what I thought and I agree! I am also leaning towards keeping this optional. I also believe that it's not for everyone, probably some Vim folks will hate the idea of an interactive form. 😄 

I've got another question. At first I was using the `octo.gh` module to build this until I realised that it's not using the API directly but rather the `gh` CLI. To avoid the callbacks I have switched to [coop.nvim](https://github.com/gregorias/coop.nvim)'s async version of `vim.system` for the `gh` CLI wrapper.
Are there any advantages of using the GitHub API (REST/GraphQL?) over the `gh` CLI? I must say I kinda prefer the idea of using the API directly but not sure whether that makes sense. Perhaps you could shine some light into how Octo works in that regard. I was struggling a little to understand its codebase since a lot of the methods seem to be generated from the API spec and in some instances the type hints are missing.

I am also using the async wrapper now in other places of my dotfiles (e.g. statusline component to display current PR). There's no direct dependency on Octo so I can continue to lazy-load it on demand.

**Comment by wd60622** [comment]
> > Not sure if forcing users to have this dependency is immediate route as it would require people to change their configs.
> 
> yep, that's what I thought and I agree! I am also leaning towards keeping this optional. I also believe that it's not for everyone – probably some Vim folks will hate the idea of an interactive form. 😄
> 
> I've got another question. At first I was using the `octo.gh` module to build this until I realised that it's not using the API directly but rather the `gh` CLI. To avoid the callbacks I have switched to [coop.nvim](https://github.com/gregorias/coop.nvim)'s async version of `vim.system` for the `gh` CLI wrapper.
> 
> Are there any advantages of using the GitHub API (REST/GraphQL?) over the `gh` CLI? I must say I kinda prefer the idea of using the API directly but not sure whether that makes sense. Perhaps you could shine some light into how Octo works in that regard. I was struggling a little to understand its codebase since a lot of the methods seem to be generated from the API spec and in some instances the type hints are missing.
> 
> I am also using the async wrapper now in other places of my dotfiles (e.g. statusline component to display current PR). There's no direct dependency on Octo so I can continue to lazy-load it on demand.

Let's discuss this in another issue
